File: packages/compose/nockerl-design-components/src/main/kotlin/com/dizyx/nockerl/design/components/NockerlCheckbox.kt
package com.dizyx.nockerl.design.components

import androidx.compose.animation.core.animateFloatAsState
import androidx.compose.animation.core.tween
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.sizeIn
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.alpha
import androidx.compose.ui.draw.clip
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.StrokeCap
import androidx.compose.ui.graphics.StrokeJoin
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.graphics.lerp
import androidx.compose.ui.semantics.Role
import androidx.compose.ui.state.ToggleableState
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import com.dizyx.nockerl.design.tokens.LocalNockerlColors
import com.dizyx.nockerl.design.tokens.NockerlMotionDuration
import com.dizyx.nockerl.design.tokens.NockerlMotionEasing
import com.dizyx.nockerl.design.tokens.NockerlProgressTrackShape
import com.dizyx.nockerl.design.tokens.nockerlLitSurface
import com.dizyx.nockerl.design.tokens.nockerlRecessedSurface
import androidx.compose.foundation.Canvas as FoundationCanvas
import androidx.compose.foundation.selection.triStateToggleable as foundationTriStateToggleable

/**
 * The standardized **checkbox**, the Tier-1 tri-state selection control,
 * porting the RATIFIED web treatments 1:1 ( / web `Checkbox.tsx@5dadb82`):
 *
 * - **Unchecked = a recessed WELL** ("fields sink"): `cardSurface3` under the
 *   shared [nockerlRecessedSurface] inner shade, ringed by the `divider` hairline.
 * - **Checked = a CONTAINED control**: the static cyan gradient
 *   (`accentPrimaryHi` → `accentPrimary`) lit from above (the shared
 *   [nockerlLitSurface] catch-light) plus a 1px **defining edge** (accent mixed
 *   68% into the shadow tint), so the filled box reads as a contained control,
 *   never a soft cyan blob.
 * - **Mixed = a DISTINCT horizontal dash** (never a faded tick), same fill.
 * - The cyan layer **cross-fades by alpha** and marks draw with on-accent ink.
 *   Fills never hard-swap (law §7).
 *
 * Square geometry (the ratified near-square 2dp track radius), 20dp box (md) /
 * 16dp (sm), and the module's shared minimum touch target regardless of box size (law §13).
 *
 * State cycling is the shared cross-platform contract [nockerlCheckboxNext]:
 * off → on, on → off, **mixed → on** (the parent tri-state resolves to all-on).
 *
 * @param state tri-state value ([ToggleableState]: On / Off / Indeterminate).
 * @param onStateChange invoked with the NEXT state per the shared contract;
 *   `null` renders a display-only box (a parent row owns the click).
 * @param modifier outer modifier.
 * @param label persistent visible label beside the box (its click target too).
 * @param description supporting line under the label.
 * @param size box scale: [NockerlCheckboxSize.MD] (default) or SM.
 * @param enabled when `false`, inert but clearly legible (off dims the well,
 *   on halves the fill, per the web's disabled ladder).
 */
@Composable
fun NockerlCheckbox(
    state: ToggleableState,
    onStateChange: ((ToggleableState) -> Unit)?,
    modifier: Modifier = Modifier,
    label: String? = null,
    description: String? = null,
    size: NockerlCheckboxSize = NockerlCheckboxSize.MD,
    enabled: Boolean = true,
) {
    val colors = LocalNockerlColors.current
    val interactive =
        if (onStateChange != null) {
            Modifier.foundationTriStateToggleable(
                state = state,
                enabled = enabled,
                role = Role.Checkbox,
                onClick = { onStateChange(nockerlCheckboxNext(state)) },
            )
        } else {
            Modifier
        }

    Row(
        modifier =
            modifier
                .then(interactive)
                .sizeIn(minWidth = MinTouchTarget, minHeight = MinTouchTarget),
        verticalAlignment = Alignment.Top,
    ) {
        CheckboxBox(state = state, size = size, enabled = enabled)
        if (label != null) {
            Column(modifier = Modifier.padding(start = 12.dp)) {
                Text(
                    text = label,
                    style = MaterialTheme.typography.bodyMedium,
                    color =
                        if (enabled) colors.onCard else colors.onCardMuted,
                )
                if (description != null) {
                    Text(
                        text = description,
                        style = MaterialTheme.typography.bodySmall,
                        color = colors.onCardMuted,
                    )
                }
            }
        }
    }
}

/** The box itself: well, cross-fading contained fill, and the drawn mark. */
@Composable
private fun CheckboxBox(
    state: ToggleableState,
    size: NockerlCheckboxSize,
    enabled: Boolean,
) {
    val colors = LocalNockerlColors.current
    val filled = state != ToggleableState.Off
    // The cyan layer CROSS-FADES by alpha over the static well (law §7); the
    // disabled-on fill sits at half strength (the web ladder), disabled-off
    // dims the whole control.
    val fillAlpha by animateFloatAsState(
        targetValue =
            when {
                !filled -> 0f
                !enabled -> 0.5f
                else -> 1f
            },
        animationSpec = tween(NockerlMotionDuration.fastMs, easing = NockerlMotionEasing.standard),
        label = "nockerlCheckboxFill",
    )
    val controlAlpha = if (!enabled && !filled) DISABLED_OFF_ALPHA else 1f
    // The 1px DEFINING EDGE: accent mixed 68% into the shadow tint (#2530a).
    val edge = lerp(colors.shadowTint, colors.accentPrimary, EDGE_ACCENT_FRACTION)
    val mark = if (enabled) colors.onAccent else colors.onAccent.copy(alpha = 0.8f)

    Box(
        modifier = Modifier.size(size.box).alpha(controlAlpha),
    ) {
        // The recessed WELL (always present under the fill, never a fill swap).
        Box(
            modifier =
                Modifier
                    .size(size.box)
                    .nockerlRecessedSurface(shape = NockerlProgressTrackShape)
                    .background(colors.cardSurface3)
                    .border(1.dp, colors.divider, NockerlProgressTrackShape),
        )
        // The contained cyan fill: static gradient + catch-light + defining edge.
        Box(
            modifier =
                Modifier
                    .size(size.box)
                    .alpha(fillAlpha)
                    .clip(NockerlProgressTrackShape)
                    .background(
                        Brush.verticalGradient(
                            colors = listOf(colors.accentPrimaryHi, colors.accentPrimary),
                        ),
                    ).nockerlLitSurface(shape = NockerlProgressTrackShape)
                    .border(1.dp, edge, NockerlProgressTrackShape),
        )
        // The MARK: tick when on, the DISTINCT dash when mixed; on-accent ink.
        FoundationCanvas(modifier = Modifier.size(size.box).alpha(fillAlpha)) {
            val w = this.size.width
            val h = this.size.height
            val stroke =
                Stroke(
                    width = size.stroke.toPx(),
                    cap = StrokeCap.Round,
                    join = StrokeJoin.Round,
                )
            when (state) {
                ToggleableState.On -> {
                    // The web tick path (M5 10.5 L8.5 14 L15 6.5 on a 20-box), scaled.
                    val p1 = Offset(w * 0.25f, h * 0.525f)
                    val p2 = Offset(w * 0.425f, h * 0.7f)
                    val p3 = Offset(w * 0.75f, h * 0.325f)
                    drawLine(mark, p1, p2, stroke.width, StrokeCap.Round)
                    drawLine(mark, p2, p3, stroke.width, StrokeCap.Round)
                }
                ToggleableState.Indeterminate -> {
                    // The distinct MIXED dash (M5.5 10 h9), scaled.
                    drawLine(
                        color = mark,
                        start = Offset(w * 0.275f, h * 0.5f),
                        end = Offset(w * 0.725f, h * 0.5f),
                        strokeWidth = stroke.width,
                        cap = StrokeCap.Round,
                    )
                }
                ToggleableState.Off -> Unit
            }
        }
    }
}

/**
 * The shared cross-platform cycle contract (identical on web + Swift):
 * off → on, on → off, and **mixed resolves to on** (the browser's native
 * indeterminate-click behavior, ratified as the family rule).
 */
fun nockerlCheckboxNext(state: ToggleableState): ToggleableState =
    when (state) {
        ToggleableState.Off -> ToggleableState.On
        ToggleableState.On -> ToggleableState.Off
        ToggleableState.Indeterminate -> ToggleableState.On
    }

/** Box scale ramp, mirroring the web `sm | md` union. */
enum class NockerlCheckboxSize(
    /** The square box edge. */
    val box: Dp,
    /** Mark stroke width. */
    val stroke: Dp,
) {
    /** 16dp box for dense rows. */
    SM(16.dp, 2.dp),

    /** 20dp box, the platform default. */
    MD(20.dp, 2.dp),
}

/** Disabled-off control opacity (the web's `.55`). */
private const val DISABLED_OFF_ALPHA = 0.55f

/** Defining-edge mix: accent 68% into shadow tint (web color-mix, #2530a). */
private const val EDGE_ACCENT_FRACTION = 0.68f

File: Sources/NockerlDesign/Components/NockerlCheckbox.swift
// NockerlCheckbox: the Tier-1 tri-state selection control, porting the RATIFIED
// web treatments 1:1 ( / web Checkbox.tsx@5dadb82):
//
//   - UNCHECKED = a recessed WELL ("fields sink"): cardSurface3 + divider hairline
//     + a top inner shade.
//   - CHECKED = a CONTAINED control: the static cyan gradient (accentPrimaryHi →
//     accentPrimary) lit from above (top catch-light) plus a 1px DEFINING EDGE
//     (accent mixed 68% into the shadow tint), so the filled box reads as a
//     contained control, never a soft cyan blob.
//   - MIXED = a DISTINCT horizontal dash (never a faded tick), same fill.
//
// The cyan layer cross-fades by opacity and the marks use on-accent ink, so fills
// never hard-swap (law §7); the cross-fade freezes under Reduce Motion. Square
// geometry (the ratified near-square 2pt track radius), 20pt box (md) / 16pt (sm).
//
// State cycling is the shared cross-platform contract `nockerlCheckboxNext`:
// off → on, on → off, mixed → on (identical on web + Compose; pure + tested).

import SwiftUI

/// Tri-state checked value (mirrors the web `boolean | 'mixed'` union).
public enum NockerlCheckedState: Equatable {
    /// Unchecked: the recessed well.
    case off
    /// Checked: the contained cyan fill + tick.
    case on
    /// Indeterminate parent: the contained fill + the DISTINCT dash.
    case mixed
}

/// The shared cycle contract (identical on web + Compose): off → on, on → off,
/// and mixed resolves to ON (the browser's native indeterminate-click rule,
/// ratified as the family behavior).
func nockerlCheckboxNext(_ state: NockerlCheckedState) -> NockerlCheckedState {
    switch state {
    case .off: return .on
    case .on: return .off
    case .mixed: return .on
    }
}

/// Box scale ramp. Mirrors the web `sm | md` union.
public enum NockerlCheckboxSize: Equatable {
    /// 16pt box for dense rows.
    case sm
    /// 20pt box, the platform default.
    case md

    /// The square box edge.
    var box: CGFloat {
        switch self {
        case .sm: return 16
        case .md: return 20
        }
    }

    /// Mark stroke width.
    var stroke: CGFloat { 2 }
}

/// The standardized tri-state checkbox (see the file header for the ratified
/// treatment). Pass `onChange: nil` for a display-only box (a parent row owns
/// the tap).
public struct NockerlCheckbox: View {
    private let state: NockerlCheckedState
    private let onChange: ((NockerlCheckedState) -> Void)?
    private let label: String?
    private let description: String?
    private let size: NockerlCheckboxSize
    private let enabled: Bool

    /// Defining-edge mix: accent 68% into shadow tint (web color-mix).
    static let edgeAccentFraction: Double = 0.68
    /// Disabled-off control opacity (the web's `.55`).
    static let disabledOffOpacity: Double = 0.55
    /// Disabled-on fill opacity (the web's `.5`).
    static let disabledOnFillOpacity: Double = 0.5

    /// Create a checkbox.
    /// - Parameters:
    ///   - state: tri-state value.
    ///   - onChange: invoked with the NEXT state per the shared contract; `nil`
    ///     renders a display-only box.
    ///   - label: persistent visible label beside the box.
    ///   - description: supporting line under the label.
    ///   - size: box scale (default md).
    ///   - enabled: when `false`, inert but clearly legible.
    public init(
        state: NockerlCheckedState,
        onChange: ((NockerlCheckedState) -> Void)? = nil,
        label: String? = nil,
        description: String? = nil,
        size: NockerlCheckboxSize = .md,
        enabled: Bool = true
    ) {
        self.state = state
        self.onChange = onChange
        self.label = label
        self.description = description
        self.size = size
        self.enabled = enabled
    }

    public var body: some View {
        let core = CheckboxBody(
            state: state, label: label, description: description,
            size: size, enabled: enabled
        )

        Group {
            if let onChange, enabled {
                Button {
                    onChange(nockerlCheckboxNext(state))
                } label: {
                    core
                }
                .buttonStyle(.plain)
            } else {
                core
            }
        }
        .accessibilityElement(children: .ignore)
        .accessibilityLabel(label ?? "Checkbox")
        .accessibilityValue(accessibleValue)
        .accessibilityAddTraits(state == .on ? [.isSelected] : [])
    }

    /// The announced tri-state value (mirrors aria-checked incl. "mixed").
    private var accessibleValue: String {
        switch state {
        case .off: return "Unchecked"
        case .on: return "Checked"
        case .mixed: return "Mixed"
        }
    }

    /// Environment-reading render body.
    private struct CheckboxBody: View {
        let state: NockerlCheckedState
        let label: String?
        let description: String?
        let size: NockerlCheckboxSize
        let enabled: Bool

        @Environment(\.colorScheme) private var colorScheme
        @Environment(\.accessibilityReduceMotion) private var reduceMotion

        var body: some View {
            let palette = NockerlPalette.resolve(colorScheme)
            let filled = state != .off
            let fillOpacity: Double = filled ? (enabled ? 1 : NockerlCheckbox.disabledOnFillOpacity) : 0
            let controlOpacity: Double = (!enabled && !filled) ? NockerlCheckbox.disabledOffOpacity : 1

            HStack(alignment: .top, spacing: NockerlSpace.space3) {
                box(palette: palette, fillOpacity: fillOpacity)
                    .frame(width: size.box, height: size.box)
                    .opacity(controlOpacity)
                    .animation(
                        reduceMotion ? nil : .nockerlStandard(NockerlMotionDuration.fast),
                        value: fillOpacity
                    )

                if let label {
                    VStack(alignment: .leading, spacing: NockerlSpace.space05) {
                        Text(label)
                            .font(.nockerl(size: NockerlFontSize.size14, weight: .light))
                            .foregroundColor(enabled ? palette.onCard : palette.onCardMuted)
                        if let description {
                            Text(description)
                                .font(.nockerl(size: NockerlFontSize.size12, weight: .light))
                                .foregroundColor(palette.onCardMuted)
                        }
                    }
                }
            }
            .contentShape(Rectangle())
        }

        /// The layered box: recessed well under a cross-fading contained fill.
        @ViewBuilder
        private func box(palette: NockerlPalette, fillOpacity: Double) -> some View {
            let shape = RoundedRectangle(cornerRadius: NockerlRadius.track, style: .continuous)
            // The 1px DEFINING EDGE: accent mixed 68% into the shadow tint.
            let edge = NockerlContrast.mix(
                palette.accentPrimary,
                into: palette.shadowTint,
                fraction: NockerlCheckbox.edgeAccentFraction
            )
            let mark = palette.onAccent

            ZStack {
                // The recessed WELL, always present under the fill.
                shape
                    .fill(palette.cardSurface3)
                    .overlay(
                        // "Fields sink": a top inner shade falling from the rim.
                        LinearGradient(
                            colors: [palette.shadowTint.opacity(0.45), .clear],
                            startPoint: .top,
                            endPoint: .center
                        )
                        .clipShape(shape)
                    )
                    .overlay(shape.strokeBorder(palette.divider, lineWidth: 1))

                // The contained cyan fill: gradient + catch-light + defining edge.
                ZStack {
                    shape.fill(
                        LinearGradient(
                            colors: [palette.accentPrimaryHi, palette.accentPrimary],
                            startPoint: .top,
                            endPoint: .bottom
                        )
                    )
                    // Top catch-light (depth law: same 1px line as every lit surface).
                    VStack(spacing: 0) {
                        palette.surfaceHighlight
                            .frame(height: 1)
                        Color.clear
                    }
                    .clipShape(shape)
                    shape.strokeBorder(edge, lineWidth: 1)

                    // The MARK: tick when on, the DISTINCT dash when mixed.
                    markPath(in: size.box)
                        .stroke(
                            mark,
                            style: StrokeStyle(
                                lineWidth: size.stroke,
                                lineCap: .round,
                                lineJoin: .round
                            )
                        )
                }
                .opacity(fillOpacity)
            }
        }

        /// The web glyph paths (tick M5 10.5 L8.5 14 L15 6.5; dash M5.5 10 h9 on a
        /// 20-box), scaled to the box edge.
        private func markPath(in edge: CGFloat) -> Path {
            var path = Path()
            switch state {
            case .on:
                path.move(to: CGPoint(x: edge * 0.25, y: edge * 0.525))
                path.addLine(to: CGPoint(x: edge * 0.425, y: edge * 0.7))
                path.addLine(to: CGPoint(x: edge * 0.75, y: edge * 0.325))
            case .mixed:
                path.move(to: CGPoint(x: edge * 0.275, y: edge * 0.5))
                path.addLine(to: CGPoint(x: edge * 0.725, y: edge * 0.5))
            case .off:
                break
            }
            return path
        }
    }
}
